### DISCUSSION TITLE
Networks with small massflow do not converge

### DISCUSSION THREAD

**Discussion by stianchris** [oemof/tespy#315] [Q&A]
I opened a similar issue some years ago, but did not further work on it. It was closed by that time.
The issue still exists: In networks with very small mass-flows (or other parameters, that are very small), newton-raphson does not converge. A possible solution would be to check for parameters, that are below a specific limit and in such cases, switch the calculation to a different unit for this parameter. In the example: switch the unit of the massflow to g/s. This needs to be done internally and not just in the unit-definition in the setup of the network. Any ideas, how this could be realized?

I added an example of a solar roof tile collector:
[tespy_issue.zip](https://github.com/oemof/tespy/files/8196504/tespy_issue.zip)
Briefly explained (schematical drawing is included): a number of solar air collectors is connected in series (in the example there are 2 solar roof tiles) and a fan is used to draw air through them. In between them, leakages (represented with valves and sources) are present, which will also draw air at ambient temperature. The mass flow at the leakages, especially of the first ones is very low. As exemplifying initialization it is set to 0.0001 kg/s (this low mass flow is normally only reached in longer series of collectors).
Two example-files are attached: one (test_srt) where first the zeta value of the valve is calculated in the design, with the given mass flow and then the mass flow should be calculated in off-design with the zeta value. For the low initial mass flow, the offdesign calculation does not converge, although all assumptions stay the same. The second test-file (test_srt_wo_srt) shows the same calculations, while zeta remains the value to be calculated and the initial leaking mass flow stays the same. This one converges without problems.

**Comment by fwitte** [comment]
Thank you for reaching out!

Since we discussed this bilaterally already, I'll just add a quick info here for everyone else:

With some smaller effort it will be possible to change unit system of mass flow to `g / s` as unit for mathematical operations. This will enable calculation of very small mass flow with the desired accuracy.

However, it seems that certain equations, i.e. the Darcy Weißbach equation or prediction pressure losses using zeta/D^4, do simply not go well with such small mass flows. Therefore, I would recommend implementing lookup tables for e.g. change in pressure vs. mass flow by using the `UserDefinedEquation` class. A similar examlpe (chang in temperature vs. volumetric flow is shown as an example in the API documentation of the class.

Have a sunny weekend :)
